### ISSUE TITLE
[User] Swap CLI

### ISSUE THREAD

**Issue by D4nte** [comit-network/xmr-btc-swap#132]
The user only needs the swap CLI to execute swap, blockchain node requirements are listed as part of the users stories.

As a user, ...:

https://github.com/comit-network/xmr-btc-swap/pull/178 - _Bitcoin wallet_:

- [x] I am not expected to run a Bitcoin full node or any Bitcoin software to execute a swap.
- [x] I need to deposit Bitcoin funds to the CLI before attempting to execute a swap.
- [ ] I can request a Bitcoin address for deposit by running the `deposit` command.
- [ ] I can request the CLI's bitcoin balance by running the `balance` command.
- [ ] I can withdraw any leftover bitcoin by using the `withdraw` command.

_Monero wallet_:

- [x] I need to run a monero-wallet-rpc instance that will be used by the CLI as a Monero Wallet.
- [x] I do not need to load my monero wallet, the swap-tool will create a temporary wallet for watching the chain, and the redeemed monero funds will be stored in another wallet that is created using `generate_from_keys`.
- [ ] Bonus: I do not have to transfer fund from the new created wallet manually, the CLI handles that for me and I do not even need to be aware of the creation of the temporary wallet. The only thing I have to provide is an address where the redeemed Monero end up.

_Config_:

- [ ] https://github.com/comit-network/xmr-btc-swap/pull/161 (needs follow up after Bitcoin wallet changes) - I can configure global parameters (ie, not specific to a given swap) using a configuration file which has reasonable default values.

_Swap_:

- [ ] I can execute a swap by running the `swap` command.
- [ ] I can resume a swap execution using `resume <swap id>` command, this will make a best effort to successfully complete the swap.
- [ ] I can check pending swaps using `list` command, do note that this command can only be ran if the CLI is stopped (and hence, no swaps are being executed).
- [ ] I can check all swaps ever made using `list --all` command.
- [ ] I do not expect for the CLI to automatically recover funds if the swap is cancelled (see [1]).
- [x] https://github.com/comit-network/xmr-btc-swap/pull/170 WIP - I can get the Bitcoin cancellation transaction and have it broadcasted by the CLI using the `cancel` command. I can force the CLI to broadcast despite expiry not being reached with `--force`. Unclear whether the CLI would then change the state of the swap at this stage. A read-only command may yield better recovery results. The CLI still prints current blockchain block and cancel expiry blockheight.
- [x] https://github.com/comit-network/xmr-btc-swap/pull/170 WIP - I can get the Bitcoin refund transaction and have it broadcasted by the CLI using the `refund` command. I can force the CLI to broadcast despite expiry not being reached with `--force`.  Unclear whether the CLI would then change the state of the swap at this stage. A read-only command may yield better recovery results. The CLI still prints current blockchain block and punish expiry blockheight.
- [ ] ~I can get the Monero refund private key, if the protocol allows it, and have it loaded in a new wallet of my `monero-wallet-rpc` using `refund` command.~ User does not refund Monero.
- [ ] Bonus: I get notified (warn log) if an expiry has passed during swap execution.

**Comment by rishflab** [comment]
> [ ] I am not expected to run a Bitcoin full node or any Bitcoin software to execute a swap.

In production this is fairly easy to resolve by replacing our `bitcoin-harness` wallet with a `bdk` wallet and using the publicly available `Esplora` API provided by Blockstream. In testing however we will have to spin up a `bitcoind` and `esplora` container to replicate the production setup on regtest

**Comment by da-kami** [comment]
Note on changes:

* I linked PRs that were put up so far on features.
* I changed the `punish` command to `refund` (this was a mistake, Bob being the user can only refund, not punish). Note, that the current implementation does not print the blockheight. @D4nte  do you think that is actually necessary?

**Comment by da-kami** [comment]
@D4nte how do you see `I do not need to load my monero wallet, I can pass the wallet name and password to the CLI that will unlock it for me.`

* The user has to pass the arguments with `--monero-wallet-name` and `--monero-wallet-password`
* The user is asked for monero-wallet-name/password upon initial startup (initial setup) and the name and password are saved to the configuration file.
* The user can specify `--load-monero-wallet` parameter and is then promtped with entering name/password upon startup. (this would have the advantage that the password is never visible/saved as cleartext - we could use https://docs.rs/dialoguer/0.7.1/dialoguer/struct.Password.html)

**Comment by da-kami** [comment]
[snip]
> I will validate that `generate_from_keys` works for Bob without having a wallet loaded initially by:
> 
>     * remove Monero wallet balance check upon startup.
> 
>     * don't load Bob's testwallet
> 
>     * execute swap on test-/stagenet and see that it goes through :)

Good that I checked 😬
Unfortunately we run into trouble with `get_height` for Bob's Monero setup. 
We record the height on Monero so we don't have to re-scan the complete blockchain when generating the new wallet. 

Generally, we might run into trouble when watching for the `transfer_proof` as well without a wallet being loaded. 
However, instead of letting the user provide a wallet (that is not used but necessary), we should actually generate a temporary wallet so we have access to the chain. 

I propose the following solution:

1. Generate temporary wallet `bob_temp` when starting up Bob (to be deleted later)
2. Since we have the wallet loaded we don't run into trouble communicating with the wallet rpc.
3. Upon redeem `generate_from_keys` will unload the temporary wallet and create the wallet where the Monero funds will be.
4. Safely remove the temporary wallet `bob_temp`

**Comment by D4nte** [comment]
> Note on changes:
> 
>     * I linked PRs that were put up so far on features.
> 
>     * I changed the `punish` command to `refund` (this was a mistake, Bob being the user can only refund, not punish). Note, that the current implementation does not print the blockheight. @D4nte  do you think that is actually necessary?

The blockheight is a nice to have.

**Comment by D4nte** [comment]
> @D4nte I am not sure if the feature `I do not need to load my monero wallet, I can pass the wallet name and password to the CLI that will unlock it for me.` makes sense for Bob given our setup, where we have a `--withdraw` command and generate a wallet by the end of the swap.
> Bob does not need to have Monero initially - we do not spend from Monero's `tx_lock` when claiming the Monero. At the moment we display the Monero balance upon startup, so a wallet connection is needed, but that might be unnecessary once the CLI is buy-only (just Bob).
> 
> So, the user (Bob) would have to start the `wallet-rpc` with `--wallet-dir` (we can provide the full command), but loading a wallet initially would not be necessary, hence no wallet-name and password required!
> 
> We'd have to validate that we can execute `generate_from_keys` against the wallet_rpc without having a wallet loaded (I am fairly certain this should work, because the open wallet just get automatically unloaded, I don't see a reason why it should fail if we don't have a wallet loaded) - then Bob actually does not have to provide any wallet information, just has to start the `wallet_rpc` in `create` mode.
> I will validate that `generate_from_keys` works for Bob without having a wallet loaded initially by:
> 
>     * remove Monero wallet balance check upon startup.
> 
>     * don't load Bob's testwallet
> 
>     * execute swap on test-/stagenet and see that it goes through :)

**Comment by D4nte** [comment]
Allow me to summarize:

From a UX point of view, we are looking at the following solution: "As a user, I can configure a Monero address to which the Monero will be sent once the transfer is complete".

Notes:
1. This will involve one more transaction and the fees associated to it.
2. We need to create few temporary wallets on `monero-wallet-rpc`, to watch the blockchain and to access the monero to redeem.
3. This actually simplifies the setup as the CLI can now starts its own `monero-wallet-rpc` on a free port, with `--wallet-dir` and freely create wallets and manipulate them.

Question: Do we currently lock the wallets we create with passwords?

**Comment by da-kami** [comment]
[snip]

> Question: Do we currently lock the wallets we create with passwords?

We pass an empty password for the wallet created with `generate_from_keys`; i.e. no password is set:

https://github.com/comit-network/xmr-btc-swap/blob/3593f5323ae09d43f8b0428c5b4e313a63a5114c/monero-harness/src/rpc/wallet.rs#L249-L257

I remember successfully testing spending from the generated wallet without providing a password.

**Comment by thomaseizinger** [comment]
Closing because epic from previous project.
